src/lib/pipe_lib/strict_pipe.ml

```diff
@@ -121,51 +131,58 @@ module Reader0 = struct
       don't_wait_for
         (let%map () = Deferred.List.iter readers ~f:Pipe.closed in
          Pipe.close_read reader.reader) ;
+      reader.downstreams <- wrap_downstreams readers;
       List.map readers ~f:wrap_reader
 
     let two reader =
       match n reader 2 with [a; b] -> (a, b) | _ -> failwith "unexpected"
   end
+
+  let rec close_downstreams = function
+    | []    -> ()
+    | r::rs -> Pipe.close_read r.reader; close_downstreams rs
 end
 
 module Writer = struct
   type ('t, 'type_, 'write_return) t =
     { type_: ('type_, 'write_return) type_
-    ; reader: 't Pipe.Reader.t
-    ; writer: 't Pipe.Writer.t }
+    ; strict_reader: 't Reader0.t
+    ; writer: 't Pipe.Writer.t
+    }
 
   (* TODO: See #1281 *)
-  let to_linear_pipe {writer= pipe; reader= _; type_= _} = pipe
+  let to_linear_pipe {writer= pipe; strict_reader= _; type_= _} = pipe
 
   let handle_overflow : type b.
       ('t, b buffered, unit) t -> 't -> b overflow_behavior -> unit =
    fun writer data overflow_behavior ->
     match overflow_behavior with
     | Crash -> raise Overflow
     | Drop_head ->
-        ignore (Pipe.read_now writer.reader) ;
+        ignore (Pipe.read_now writer.strict_reader.reader) ;
         Pipe.write_without_pushback writer.writer data
 
   let write : type type_ return. ('t, type_, return) t -> 't -> return =
    fun writer data ->
     match writer.type_ with
     | Synchronous -> Pipe.write writer.writer data
     | Buffered (`Capacity capacity, `Overflow overflow) ->
-        if Pipe.length writer.reader > capacity then
+        if Pipe.length writer.strict_reader.reader > capacity then
           handle_overflow writer data overflow
         else Pipe.write_without_pushback writer.writer data
 
-  let close {writer; _} = Pipe.close writer
+  let close {type_ = _; strict_reader; writer} =
+    Pipe.close writer;
+    Reader0.close_downstreams strict_reader.downstreams
 
   let is_closed {writer; _} = Pipe.is_closed writer
 end
 
 let create type_ =
   let reader, writer = Pipe.create () in
-  let reader, writer =
-    (Reader0.{reader; has_reader= false}, Writer.{type_; reader; writer})
-  in
-  (reader, writer)
+  let strict_reader = Reader0.{reader; has_reader= false; downstreams= []} in
+  let strict_writer = Writer.{type_; strict_reader; writer}
+  in (strict_reader, strict_writer)
 
 let transfer reader {Writer.writer; _} ~f =
```

Comment:
add the `_` (which you'll now pattern match on) as a downstream reader for the first parameter `reader`